### PR TITLE
Graph Visualization: Added a 'Label' structure, debug marks , order of commandBuffer and reestructuration of the code.

### PR DESCRIPTION
**Pull request by ecapia**

### CODE REVIEW

File: gapis/resolve/dependencygraph2/graph_visualization/graph_algorithms.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+)
+
+const (
+	NO_VISITED       = 0
+	VISITED_AND_USED = -1
+	FRAME            = "FRAME"
+	UNUSED           = "UNUSED"
+)
+
+func (g *graph) traverseGraph(currentNode *node, visitTime, minVisitTime, idInStronglyConnectedComponents, visitedNodesId *[]int, currentId, currentTime *int) {
```
Comment (AWoloszyn) [inline gapis/resolve/dependencygraph2/graph_visualization/graph_algorithms.go]: These are large enough functions, it would be good to get comments on exactly what the parameters are supposed to be.

File: gapis/api/graph_visualization.go [hunk 2]
```diff
@@ -14,6 +14,72 @@
...
+	"bytes"
+	"fmt"
+)
+
+type Label struct {
+	LevelsName []string
+	LevelsID   []int
+}
+
+func (label *Label) GetSize() int {
```
Comment (AWoloszyn) [inline gapis/api/graph_visualization.go]: Godocs needed for all public methods.

File: gapis/api/graph_visualization.go [hunk 3]
```diff
@@ -14,6 +14,72 @@
 
 package api
 
+import (
+	"bytes"
+	"fmt"
+)
+
```
Comment (AWoloszyn) [inline gapis/api/graph_visualization.go]: Godocs here as well.

File: gapis/api/vulkan/graph_visualization.go [hunk 4]
```diff
@@ -66,60 +75,127 @@ func getSubCommandHierarchyNames() *api.HierarchyNames {
...
 		}
 	}
-	if _, ok := builder.labelToHierarchy[label]; !ok {
-		builder.labelToHierarchy[label] = &api.Hierarchy{}
+	temporalLabelAsAString := label.GetLabelAsAString()
+	if _, ok := builder.labelAsAStringToHierarchy[temporalLabelAsAString]; !ok {
+		builder.labelAsAStringToHierarchy[temporalLabelAsAString] = &api.Hierarchy{}
 	}
-	hierarchy := builder.labelToHierarchy[label]
+	hierarchy := builder.labelAsAStringToHierarchy[temporalLabelAsAString]
```
Comment (bjoeris) [inline gapis/api/vulkan/graph_visualization.go]: In general this pattern of inserting a default value into the map when the key is missing can be re-expressed as
```
hierarchy, ok := builder.labelAsAStringToHierarchy[temporalLabelAsAString]
if !ok {
	hierarchy = &api.Hierarchy{}
	builder.labelAsAStringToHierarchy[temporalLabelAsAString] = hierarchy
}		
```
This removes one of the map lookups.

File: gapis/resolve/dependencygraph2/graph_visualization/graph_structure.go [hunk 5]
```diff
@@ -197,122 +189,6 @@ func (input nodeSorter) Less(i, j int) bool {
 	return input[i].id < input[j].id
 }
 
-func (g *graph) traverseGraph(currentNode *node, visitTime, minVisitTime, idInStronglyConnectedComponents, visitedNodesId *[]int, currentId, currentTime *int) {
```
Comment (bjoeris) [inline gapis/resolve/dependencygraph2/graph_visualization/graph_structure.go]: When moving big chunks of code like this, it's usually preferred to do the move in a separate commit, since it is hard to see what changed. Just a comment, no need to change this.

File: gapis/api/vulkan/graph_visualization.go [hunk 6]
```diff
@@ -66,60 +75,127 @@ func getSubCommandHierarchyNames() *api.HierarchyNames {
...
+				builder.labelsInsideDebugMarkers = append(builder.labelsInsideDebugMarkers, currentLabel)
+				builder.positionOfDebugMarkersBegin = append(builder.positionOfDebugMarkersBegin, len(builder.labelsInsideDebugMarkers)-1)
+				addDebugMarker(builder, positionOfLastDebugMarkerBegin, len(builder.labelsInsideDebugMarkers)-1)
+				builder.positionOfDebugMarkersBegin = builder.positionOfDebugMarkersBegin[:len(builder.positionOfDebugMarkersBegin)-1]
+			}
+		}
+
+	} else if len(builder.positionOfDebugMarkersBegin) > 0 {
+		if labelOfLastDebugMarkerBegin.GetSize() != currentLabel.GetSize() {
+			builder.positionOfDebugMarkersBegin = builder.positionOfDebugMarkersBegin[:0]
```
[the remark this answers is not in the document]
Comment (AWoloszyn) [reply gapis/api/vulkan/graph_visualization.go]: We have a ton of code in vulkan.go to handle nesting of debug markers and similar, (as used for displaying groups in the UI) It handles some of these cases (bad nestings etc).

Is there a chance we can leverage that? 
It may be a follow-up, but it may save some time in handling some of these cases.

File: gapis/resolve/dependencygraph2/graph_visualization/graph_algorithms.go (hunk 1 above)
Comment (ecapia) [reply gapis/resolve/dependencygraph2/graph_visualization/graph_algorithms.go]: I abstracted the parameters used to compute the Strongly Connected Components i a directed graph based on Tarjan algorithm. Thank you.

File: gapis/api/vulkan/graph_visualization.go (hunk 4 above)
Comment (ecapia) [reply gapis/api/vulkan/graph_visualization.go]: Thank you. Yes , using this I am removing one look up in the map. I changed it for the two cases.

File: gapis/api/vulkan/graph_visualization.go [hunk 7]
```diff
@@ -66,60 +75,127 @@ func getSubCommandHierarchyNames() *api.HierarchyNames {
...
+				label.PushBackLabel(hierarchyLabel)
 			} else {
-				label += fmt.Sprintf("/%d", index[i])
+				label.PushBack("", int(index[i]))
 			}
 		}
 	}
-	if _, ok := builder.labelToHierarchy[label]; !ok {
-		builder.labelToHierarchy[label] = &api.Hierarchy{}
+	temporalLabelAsAString := label.GetLabelAsAString()
```
[the remark this answers is not in the document]
Comment (ecapia) [reply gapis/api/vulkan/graph_visualization.go]: It is intentional, but I figured out that 'temporal' is not the correct word to describe something provisional , thank you. I changed it to temporary.

File: gapis/api/vulkan/graph_visualization.go [hunk 8]
```diff
@@ -66,60 +75,127 @@ func getSubCommandHierarchyNames() *api.HierarchyNames {
...
+		}
+	}
+	return size
+}
+
+func addDebugMarker(builder *labelForVulkanCommands, from, to int) {
+	level := builder.labelsInsideDebugMarkers[len(builder.labelsInsideDebugMarkers)-1].GetSize() - 1
+	builder.numberOfDebugMarker++
+	for i := from; i <= to; i++ {
+		builder.labelsInsideDebugMarkers[i].Insert(level, VK_CMD_DEBUG_MARKER, builder.numberOfDebugMarker)
```
[the remark this answers is not in the document]
Comment (ecapia) [reply gapis/api/vulkan/graph_visualization.go]: In fact some labels could be affected by more than one debug marker (nested debug markers), in that case the algorithm add twice the debug marker level.

File: gapis/api/vulkan/graph_visualization.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ecapia) [reply gapis/api/vulkan/graph_visualization.go]: Now I am adding debug markers as long as it does not destroy any level of hierarchy defined in commands. So, if exists some command between debug_marker_begin and debug_marker_end which has smaller Label size , adding the debug marker here will destroy the levels of hierarchy and for this case I just ignore the debug marker. It means I am preserving Labels with no-decreasing size. I have tested with several captures and it is working very nice. In addition to this, I am analizing the possibility to include code from vulkan.go to addres this problem as well, but if it is the case , I will add this in the last commit.

File: gapis/api/graph_visualization.go (hunk 2 above)
Comment (ecapia) [reply gapis/api/graph_visualization.go]: Thank you, I added the Godocs for all public methods.

File: gapis/resolve/dependencygraph2/graph_visualization/graph_structure.go (hunk 5 above)
Comment (ecapia) [reply gapis/resolve/dependencygraph2/graph_visualization/graph_structure.go]: I have abstracted this part of code to simplify the understanding, because it is essentially the Tarjan algorithm to compute Strongly Connected Components,

File: gapis/api/vulkan/graph_visualization.go [hunk 9]
```diff
@@ -66,60 +75,129 @@ func getSubCommandHierarchyNames() *api.HierarchyNames {
...
-			return commandBuffer
+			return parameter.Get().(VkCommandBuffer), true
 		}
 	}
-	return ""
+	return 0, false
 }
 
-func (builder *labelForVulkanCommands) GetCommandLabel(command api.Cmd, commandNodeId uint64) string {
+func (builder *labelForVulkanCommands) GetCommandLabel(command api.Cmd, cmdId uint64) *api.Label {
```
Comment (bjoeris) [inline gapis/api/vulkan/graph_visualization.go]: These functions that implement the interfaces are also public, and should also have godocs. Those godocs are usually exactly the same as the godocs from the interface method they are implementing.

Comment (ecapia) [reply gapis/api/vulkan/graph_visualization.go]: Thank you, I will add the Godocs here as well.
